File: backend/controlplane/internal/httpapi/special_formats.go
package httpapi

import (
	"os"
	"path/filepath"
	"strings"
)

// Special-format registry (Go side). Mirrors backend/shared/special_formats.json and
// the Python registry (imaging/special_formats.py); special_formats_parity_test.go
// asserts this matches the shared JSON. These are formats that are directory/archive
// shaped or need a non-default serving backend.
//
// TEMPLATE — to add a special format: append a row here, mirror it in the shared JSON +
// the Python registry, and wire its serve backend. Detection is by extension (incl.
// series-suffixed names like "scan.zarr_3") OR, for directory formats, by marker files.
type specialFormat struct {
	ID           string
	Extensions   []string // lowercase, with leading dot
	Shape        string   // "file" | "directory" | "archive"
	DirMarkers   []string // for directory formats: files that must be present at the root
	Serve        string   // "ngff" | "libbioimage" | "transcode"
	ResourceKind string   // "image" | ...
}

var specialFormats = []specialFormat{
	{
		ID:           "ome-zarr",
		Extensions:   []string{".ome.zarr", ".zarr"},
		Shape:        "directory",
		DirMarkers:   []string{".zattrs", ".zgroup"},
		Serve:        "ngff",
		ResourceKind: "image",
	},
}

// detectSpecialFormatByName matches a resource's name against the registry's extensions,
// tolerating a series/page suffix (e.g. "img.zarr_3").
func detectSpecialFormatByName(name string) *specialFormat {
	lower := strings.ToLower(strings.TrimSpace(name))
	for i := range specialFormats {
		for _, ext := range specialFormats[i].Extensions {
			if strings.HasSuffix(lower, ext) || strings.Contains(lower, ext+"_") {
				return &specialFormats[i]
			}
		}
	}
	return nil
}

// detectSpecialFormatByDir matches a directory against directory-format marker files
// (e.g. an OME-Zarr group has .zattrs + .zgroup at its root).
func detectSpecialFormatByDir(path string) *specialFormat {
	info, err := os.Stat(path)
	if err != nil || !info.IsDir() {
		return nil
	}
	for i := range specialFormats {
		sf := &specialFormats[i]
		if sf.Shape != "directory" || len(sf.DirMarkers) == 0 {
			continue
		}
		all := true
		for _, marker := range sf.DirMarkers {
			if _, err := os.Stat(filepath.Join(path, marker)); err != nil {
				all = false
				break
			}
		}
		if all {
			return sf
		}
	}
	return nil
}

// specialFormatForResource resolves the special format of a resource by name first, then
// (for unresolved names) by probing the on-disk directory markers.
func specialFormatForResource(record resourceRecord, path string) *specialFormat {
	if sf := detectSpecialFormatByName(record.OriginalName); sf != nil {
		return sf
	}
	return detectSpecialFormatByDir(path)
}

// servesViaNgff reports whether a resource should be served by the ngff-service.
func (deps ServerDeps) servesViaNgff(record resourceRecord, path string) bool {
	if strings.TrimSpace(deps.NgffServiceURL) == "" {
		return false
	}
	sf := specialFormatForResource(record, path)
	return sf != nil && sf.Serve == "ngff"
}

// ngffDeps returns a copy of deps with ImageServiceURL pointed at the ngff-service, so
// the existing proxy helpers (cache, backpressure, fallback) route to it unchanged.
func (deps ServerDeps) ngffDeps() ServerDeps {
	d := deps
	d.ImageServiceURL = deps.NgffServiceURL
	return d
}

File: backend/deepagents_runtime/src/ultra_deepagents/imaging/special_formats.py
"""Special-format registry (Python side).

Mirrors ``backend/shared/special_formats.json`` and the Go registry
(``internal/httpapi/special_formats.go``); ``tests/test_special_formats_parity.py``
asserts this matches the shared JSON. Special formats are directory/archive shaped or
need a non-default serving backend (e.g. OME-Zarr → the ngff-service).

TEMPLATE — to add a special format: append a row to the shared JSON, mirror it here and
in the Go registry, and wire its reader/serve hook. The parity test enforces the mirror.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass

__all__ = ["SpecialFormat", "SPECIAL_FORMATS", "detect_by_name", "detect_by_dir", "detect"]


@dataclass(frozen=True)
class SpecialFormat:
    id: str
    extensions: tuple[str, ...]  # lowercase, with leading dot
    shape: str  # "file" | "directory" | "archive"
    dir_markers: tuple[str, ...]
    serve: str  # "ngff" | "libbioimage" | "transcode"
    reader: str
    resource_kind: str


SPECIAL_FORMATS: list[SpecialFormat] = [
    SpecialFormat(
        id="ome-zarr",
        extensions=(".ome.zarr", ".zarr"),
        shape="directory",
        dir_markers=(".zattrs", ".zgroup"),
        serve="ngff",
        reader="ngff",
        resource_kind="image",
    ),
]


def detect_by_name(name: str) -> SpecialFormat | None:
    """Match a resource name against the registry extensions (tolerating a series/page
    suffix like ``img.zarr_3``)."""
    lower = (name or "").strip().lower()
    for sf in SPECIAL_FORMATS:
        for ext in sf.extensions:
            if lower.endswith(ext) or (ext + "_") in lower or re.search(re.escape(ext) + r"_\d+$", lower):
                return sf
    return None


def detect_by_dir(path: str) -> SpecialFormat | None:
    """Match a directory against directory-format marker files (e.g. an OME-Zarr group
    has .zattrs + .zgroup at its root)."""
    if not os.path.isdir(path):
        return None
    for sf in SPECIAL_FORMATS:
        if sf.shape != "directory" or not sf.dir_markers:
            continue
        if all(os.path.exists(os.path.join(path, m)) for m in sf.dir_markers):
            return sf
    return None


def detect(name: str, path: str | None = None) -> SpecialFormat | None:
    """Resolve by name first, then (if a path is given) by on-disk directory markers."""
    sf = detect_by_name(name)
    if sf is not None:
        return sf
    return detect_by_dir(path) if path else None
